synfig-studio/src/gui/workarearenderer/renderer_canvas.cpp

```diff
@@ -371,19 +372,91 @@ Renderer_Canvas::build_onion_frames()
 		const Color color_future(0.f, 1.f, 0.f, 0.2f);
 		const ColorReal base_alpha = 1.f;
 		const ColorReal current_alpha = 0.5f;
-		// make onion levels
-		for(int i = past; i > 0; --i) {
-			Time time = base_time - frame_duration*i;
-			ColorReal alpha = base_alpha + (ColorReal)(past - i + 1)/(ColorReal)(past + 1);
-			if (time >= rend_desc.get_time_start() && time <= rend_desc.get_time_end())
-				onion_frames.push_back(FrameDesc(time, w, h, alpha));
+
+		if (keyframes)
+		{
+			// Onion Skin on Keyframes
+			Time keyframe_time;
+
+			int n_past_keyframes = 0;
+			int n_future_keyframes = 0;
+
+			// Count available past Keyframes to correctly calculate alpha value
+			keyframe_time = base_time;
+			for(int i = past; i > 0; --i) {
+				KeyframeList::iterator iter;
+				if (canvas->keyframe_list().find_prev(keyframe_time, iter)) {
+					n_past_keyframes++;
+					keyframe_time = iter->get_time();
+				}
+				else
+					break;
+			}
```

Comment:
I suggest to change it a bit:
you store in a std::vector<Time> the keyframe valid times (including if it is in the RendDesc time range). One vector for past and other for future skins.

You can do the same for non-keyframe onion skin.

Then later you just iterate over these vector more easily and clean to read.

Reply:
Followed you suggestion, applied in the last commit